gnomad/utils/constraint.py

```diff
@@ -360,3 +412,285 @@ def collapse_strand(
+        agg_expr[f"observed_{pop}"] = hl.agg.array_sum(
+            high_cov_ht[f"downsampling_counts_{pop}"]
+        )
+
+    # Generate a Table with all necessary annotations (x and y listed above)
+    # for the plateau models.
+    high_cov_group_ht = high_cov_ht.group_by(*keys).aggregate(**agg_expr)
+    high_cov_group_ht = annotate_mutation_type(high_cov_group_ht)
+
+    # Build plateau models.
+    plateau_models_agg_expr = build_plateau_models(
+        cpg_expr=high_cov_group_ht.cpg,
+        mu_snp_expr=high_cov_group_ht.mu_snp,
+        observed_variants_expr=high_cov_group_ht.observed_variants,
+        possible_variants_expr=high_cov_group_ht.possible_variants,
+        pop_observed_variants_exprs_arr=hl.array(
+            [high_cov_group_ht[f"observed_{pop}"] for pop in pops]
+        ),
+        weighted=weighted,
+    )
+    plateau_models = dict(
+        high_cov_group_ht.aggregate(hl.struct(**plateau_models_agg_expr))
+    )
+
+    if pops:
+        # Map the models to their corresponding populations.
+        pop_models_arr = plateau_models["pop"]
+        for idx, pop in enumerate(pops):
+            plateau_models[pop] = pop_models_arr[idx]
+        plateau_models.pop("pop")
+    plateau_models = hl.struct(**plateau_models)
+
+    # Filter to sites with coverage below `cov_cutoff` and larger than 0.
+    low_cov_ht = coverage_ht.filter(
+        (coverage_ht.exome_coverage < cov_cutoff) & (coverage_ht.exome_coverage > 0)
+    )
+
+    # Metric that represents the relative mutability of the exome calculated on high
+    # coverage sites and will be used as scaling factor when building the coverage
+    # model.
+    high_coverage_scale_factor = high_cov_ht.aggregate(
+        hl.agg.sum(high_cov_ht.observed_variants)
+        / hl.agg.sum(high_cov_ht.possible_variants * high_cov_ht.mu_snp)
+    )
+
+    # Generate a Table with all necessary annotations (x and y listed above)
+    # for the coverage model.
+    low_cov_group_ht = low_cov_ht.group_by(
+        log_coverage=hl.log10(low_cov_ht.exome_coverage)
+    ).aggregate(
+        low_coverage_oe=hl.agg.sum(low_cov_ht.observed_variants)
+        / (
+            high_coverage_scale_factor
+            * hl.agg.sum(low_cov_ht.possible_variants * low_cov_ht.mu_snp)
+        )
+    )
+
+    # Build the coverage model.
+    # TODO: consider weighting here as well
+    coverage_model_expr = build_coverage_model(
+        low_coverage_oe_expr=low_cov_group_ht.low_coverage_oe,
+        log_coverage_expr=low_cov_group_ht.log_coverage,
+    )
+    coverage_model = tuple(low_cov_group_ht.aggregate(coverage_model_expr).beta)
+    return coverage_model, plateau_models
+
+
+def build_plateau_models(
+    cpg_expr: hl.expr.BooleanExpression,
+    mu_snp_expr: hl.expr.Float64Expression,
+    observed_variants_expr: hl.expr.Int64Expression,
+    possible_variants_expr: hl.expr.Int64Expression,
+    pop_observed_variants_exprs_arr: Dict[str, hl.ArrayNumericExpression] = None,
```

Comment:
```suggestion
    pop_observed_variants_array_expr: hl.expr.ArrayExpression = None,
```
updated name and type